### PR TITLE
[3743] Adds support for user search based on login or email

### PR DESCRIPTION
**Pull request by dpavlos**
- Adds support for #3743 . I have used the already implemented `GetUserByLogin` in `sqlstore/user.go` to achieve this. The returned data is modelled upon the `UserProfileDTO` in order to keep consistency with the `GetUserById` call.
- Adds documentation for the new call.
- Adds correct http response header in `GetUserById` in case of the user is not found.
- Minor typo fixes.

### CODE REVIEW

File: docs/sources/http_api/user.md [hunk 1]
```diff
@@ -69,6 +69,38 @@ parent = "http_api"
       "isGrafanaAdmin": true
     }
 
+## Get single user by Username(login) or Email
+
+    `GET /api/users/loginoremail/:loginOrEmail`
+
+    **Example Request using the email option**:
+
+        GET /api/loginoremail/user@mygraf.com HTTP/1.1
```
Comment (torkelo) [inline docs/sources/http_api/user.md]: the example url looks wrong here. 

I think a better api url / action would be   `/api/users/lookup?loginOrEmail=mylogin`   that is use a query parameter, and a `lookup` URL resource

Comment (dpavlos) [inline docs/sources/http_api/user.md]: Thanks @torkelo. Indeed the lookup url looks better. I'll have a look into it and get back to you soon.

File: pkg/api/user.go [hunk 2]
```diff
@@ -33,7 +33,7 @@ func getUserUserProfile(userId int64) Response {
 
 // GET /api/users/loginoremail/:loginOrEmail
 func GetUserByLoginOrEmail(c *middleware.Context) Response {
-	query := m.GetUserByLoginQuery{LoginOrEmail: c.Params(":loginOrEmail")}
+	query := m.GetUserByLoginQuery{LoginOrEmail: c.QueryEscape("loginOrEmail")}
```
Comment (torkelo) [inline pkg/api/user.go]: not sure QueryEscape is the right method to use here, it HTML encodes the email, the email is not stored html encoded

Comment (dpavlos) [inline pkg/api/user.go]: For simple cases it returns correct results, however I agree that it is not the right one to be used here. On the other hand macaron doesn't provide a simple func for url string queries. The only option AFAICS which is close to what we are trying to achieve is the Query() which takes also into account post requests for forms. Use directly the `c.Req.URL.Query().Get("loginOrEmail")` might be a good and simple option though.

I have two more concerns regarding the lookup and email query that would like to discuss a bit further.
1) How do you handle the RFC compliance for emails on such cases? For example if a user provide an email address which contains special characters in local-part.
2) Security-wise it is a bit of a bad practice to pass personal information (in this case the email address) via URL with a GET request. Perhaps a more suitable solution would be to use POST instead and pass the parameter via the body. However this doesn't follow very well the REST guidelines.

Comment (torkelo) [inline pkg/api/user.go]: > For simple cases it returns correct results 

What case does it not return correct results? 

> which takes also into account post requests for forms. 

But this is a GET request so not sure that applies here. 

1) special charachers are URL encoded by user who uses the API, decoded automatically by golang http library
2) Grafana is not passing any personal information as this is an external API call (That grafana UI is not using). And there is no inherent security difference between POST and GET if the call is unencrypted

Comment (dpavlos) [inline pkg/api/user.go]: Thanks for the reply @torkelo . 
> What case does it not return correct results?
I had in mind the email tagging technique (there might be other more complicated use cases) which uses the '+' character in local-part but I double checked it with all the available functions we've discussed so far and it seems to work (with the appropriate encoding). 
> But this is a GET request so not sure that applies here.
I was meant to say that it has an extra logic we don't need/use in this case. But I agree that it is safer not to use the `QueryEscape()`. I have updated the PR to use the `Query()` instead. Let me know if you agree. :)

2. Yeah.. I had in mind https and browser history (parameters will be present in the browser history when inserted via the address bar) when I wrote this. Apparently it is not the case at the moment.

**Comment by vvgelder** [comment]
I would use this function GetUserByLoginOrEmail to retrieve the user id, but I noticed it doesn't return the id. If I already knew the user id, I would use GetUserByID, but if I don't know the id I would expect to get the id returned by this function so I can use that in other api calls that require the user id. Wouldn't it be better to use UserSearchHitDTO as return value?

**Pull request merged**